src/components/input/input.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * Grow Elastic TextArea
+ *
+ * Increases the `rows` attribute of a `textarea` until it no longer scrolls.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const growElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Grow until we stop scrolling
+  while (isScrolling) {
+    rows += 1;
+    textarea.setAttribute('rows', String(rows));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+  }
+};
+
+/**
+ * Shrink Elastic TextArea
+ *
+ * Decreases the `rows` attribute of a `textarea` until it begins scrolling.
+ *
+ * @param {HTMLTextAreaElement} textarea - the target `textarea` element
+ */
+const shrinkElasticTextArea = (textarea: HTMLTextAreaElement) => {
+  const minRows = Number(textarea.dataset.minRows);
+  let rows = Number(textarea.getAttribute('rows')) || minRows;
+  let isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+  // Shrink until we hit the minimum rows or start scrolling
+  for (let i = rows; i >= minRows; i--) {
+    rows = i;
+    textarea.setAttribute('rows', String(Math.max(rows, minRows)));
+    isScrolling = textarea.scrollHeight > textarea.clientHeight;
+
+    // If we're scrolling, then we shrank too much, so grow a bit and stop
+    if (isScrolling) {
+      growElasticTextArea(textarea);
+      break;
+    }
+  }
+};
+
+/**
+ * Update Elastic TextArea
+ *
+ * When a textarea is updated, compares the new value to the previous value
+ * (from a dataset attribute) to determine whether to grow or shrink.
+ *
+ * Note: Ideally we'd destructure `event.target` from the parameters and declare
+ * the type directly, since we don't care about the rest of the `event` object.
+ * Unfortunately, that caused some TS problems. If someone finds a solution,
+ * we should update this function to take `{target}` as the only parameter.
+ * @see https://stackoverflow.com/a/28900856/4898045
+ *
+ * @param {Event} event - the input event on the target `textarea`
+ */
+const updateElasticTextArea = (event: Event) => {
+  const textarea = event.target as HTMLTextAreaElement;
+  const lastValue = textarea.dataset.lastValue || '';
+  const newValue = textarea.value;
+
+  // Check whether to grow or shrink
+  if (newValue.length > lastValue.length) {
+    growElasticTextArea(textarea);
+  } else {
+    shrinkElasticTextArea(textarea);
+  }
```

Comment:
I'm not sure this is the best way to tell whether we need to grow or shrink. e.g. Take these two values:

**lastValue:**

```
I am a single line but I have lots of characters! I only need one row since I have no returns!
```

**newValue:**
```
I
need
more
rows!
```

In this case the new value is shorter than the old value, but the textarea needs to grow! This probably wouldn't happen if I was typing away, but it could happen if I copy/paste replaced my old text.

Could we use a similar strategy to what we did for `isScrolling`? (I realize this logic is probably carried over from the old implementation)

Reply:
I realize this is an edge case and not a regression so I'd be fine filing an issue and moving on if you'd like

Reply:
I can reproduce this issue in-browser if I paste in the first text, then select all and paste in the second text and then undo-redo. The single-line one does not make the text box resize smaller

Reply:
Heh. Tricky edge case!

The shrink method is clever enough to handle this — it'll shrink until it either hits the minimum rows value or starts scrolling, and then it'll grow until it stops scrolling.

But the grow method is dumb, and just adds rows as long as it's scrolling. It doesn't the same sanity check as shrink does.

I don't know how to fix it, though — grow would need to check whether the content is currently _shorter_ than the height. And I don't know how to check that. Does anyone have any ideas?

(Also, as Paul points out, this edge case is present in the original component as well, so if this is hairy to fix, perhaps this should be a new issue, rather than a part of this PR)

Reply:
How about this:

Currently it works like this:
```
If the contents.length is longer then before, grow()
Else shrink()
```

How about we change it to:
```
If there is a scrollbar, grow()
Else shrink()
```

I think that would fix this edge case

Reply:
Clever, and it totally works!